lighthouse-cli/chrome-launcher.js

```diff
@@ -0,0 +1,220 @@
+const execSync = childProcess.execSync;
+const spawnSync = childProcess.spawnSync;
+
+module.exports = class Launcher {
+  constructor(opts) {
+    opts = opts || {};
+    // choose the first one (default)
+    this.head = defaults(opts.head, true);
+    this.pollInterval = 500;
+    this.chrome = null;
+    this.prepared = false;
+  }
+
+  flags() {
+    const flags = [
+      '--remote-debugging-port=9222',
+      '--no-first-run',
+      `--user-data-dir=${this.TMP_PROFILE_DIR}`
+    ];
+
+    if (process.platform === 'linux') {
+      flags.push('--disable-setuid-sandbox');
+    }
+
+    return flags;
+  }
+
+  prepare() {
+    switch (process.platform) {
+      case 'darwin':
+      case 'linux':
+        this.TMP_PROFILE_DIR = unixTmpDir();
+        break;
+
+      case 'win32':
+        this.TMP_PROFILE_DIR = win32TmpDir();
+        break;
+
+      default:
+        throw new Error('Platform ' + process.platform + ' is not supported');
+    }
+
+    this.outFile = fs.openSync(`${this.TMP_PROFILE_DIR}/chrome-out.log`, 'a');
+    this.errFile = fs.openSync(`${this.TMP_PROFILE_DIR}/chrome-err.log`, 'a');
+    this.pidFile = fs.openSync(`${this.TMP_PROFILE_DIR}/chrome.pid`, 'w');
+
+    console.log(`created ${this.TMP_PROFILE_DIR}`);
+
+    this.prepared = true;
+  }
+
+  run() {
+    if (!this.prepared) {
+      this.prepare();
+    }
+
+    return Promise.resolve()
+      .then(() => {
+        const installations = chromeFinder[process.platform]();
+
+        switch (true) {
+          case installations.length < 1:
+            return Promise.reject(new Error('No Chrome Installations Found'));
+
+          case installations.length === 1:
+            return installations[0];
+
+          case this.head:
+            return installations[0];
+
+          default:
+            return ask('Choose a Chrome installation to use with Lighthouse', installations);
+        }
+      })
+      .then(execPath => this.spawn(execPath));
+  }
+
+  spawn(execPath) {
+    return new Promise((resolve, reject) => {
+      const chrome = spawn(
+        execPath,
+        this.flags(),
+        {
+          detached: true,
+          stdio: ['ignore', this.outFile, this.errFile]
+        }
+      );
+      this.chrome = chrome;
+
+      fs.writeFileSync(this.pidFile, chrome.pid.toString());
+
+      console.log('Chrome running with pid =', chrome.pid);
+      resolve(chrome.pid);
+    })
+    .then(pid => Promise.all([pid, this.poll(0)]));
+  }
+
+  cleanup(client) {
+    if (client) {
+      client.removeAllListeners();
+      client.end();
+      client.destroy();
+      client.unref();
+    }
+  }
+
+  connect() {
+    return new Promise((resolve, reject) => {
+      const client = net.createConnection(9222);
+      client.once('error', err => {
+        this.cleanup(client);
+        reject(err);
+      });
+      client.once('connect', _ => {
+        this.cleanup(client);
+        resolve();
+      });
+    });
+  }
+
+  poll() {
```

Comment:
So, it could be `isDebuggerReady` and `waitUntilDebuggerReady` ... but it's too long ... So I thought connect and poll would be better because that's what they do ...